src/main/java/org/opensearch/sdk/sample/crud/rest/CrudRestHandler.java

```diff
@@ -0,0 +1,108 @@
+        super();
+        // initializes pathTrie
+        routes();
+    }
+
+    @Override
+    public List<Route> routes() {
+        List<Route> routes = List.of(
+                new ProtectedRoute(GET, "/detector", "ListDetector", (Method method, String uri) -> new ExtensionRestResponse(OK, CREATE_SUCCESS, List.of())),
+                new ProtectedRoute(PUT, "/detector", "CreateDetector", new CrudCreateRestHandler()),
+                new ProtectedRoute(POST, "/detector/{detector_id}", "UpdateDetector", new CrudUpdateRestHandler()),
+                new ProtectedRoute(GET, "/detector/{detector_id}/results", "ListResults", (Method method, String uri) -> new ExtensionRestResponse(OK, CREATE_SUCCESS, List.of())),
+                new ProtectedRoute(GET, "/detector/{detector_id}/results/{results_id}", "GetResults",(Method method, String uri) -> new ExtensionRestResponse(OK, CREATE_SUCCESS, List.of())),
+                new ProtectedRoute(DELETE, "/detector/{detector_id}/results/{results_id}", "DeleteResults", (Method method, String uri) -> new ExtensionRestResponse(OK, CREATE_SUCCESS, List.of()))
+        );
+        // Only initialize this on first call
+        if (pathTrie == null) {
+            pathTrie = new PathTrie<>(RestUtils.REST_DECODER);
+            for (Route r : routes) {
+                List<Route> routesForPath = pathTrie.retrieve(r.getPath());
+                if (routesForPath == null) {
+                    routesForPath = new ArrayList<>();
+                    routesForPath.add(r);
+                    pathTrie.insert(r.getPath(), routesForPath);
+                } else {
+                    routesForPath.add(r);
+                }
+            }
+        }
+        return routes;
+    }
+
+    // How should the extension list what permissions it wants to create?
+    // Will the permissions be part of the API spec of the extension?
+    @Override
+    public ExtensionRestResponse handleRequest(Method method, String uri) {
+        if (pathTrie.retrieve(uri) == null || pathTrie.retrieve(uri).isEmpty()) {
+            return new ExtensionRestResponse(
+                NOT_FOUND,
+                "Extension REST action improperly configured to handle " + method.name() + " " + uri,
+                List.of()
+            );
+        }
+        ProtectedRoute matchingRoute = getMatchingRoute(method, uri);
+        if (matchingRoute != null) {
+            return matchingRoute.handleRequest(method, uri);
+        }
+        return new ExtensionRestResponse(
+                NOT_FOUND,
+                "Extension REST action improperly configured to handle " + method.name() + " " + uri,
+                List.of()
+        );
+    }
+
+    public ProtectedRoute getMatchingRoute(Method method, String uri) {
+        for (Route r : routes()) {
+            ProtectedRoute protectedRoute = (ProtectedRoute) r;
+            if (protectedRoute.getMethod() == method && uri.matches(protectedRoute.getRouteRegex().pattern())) {
```

Comment:
this iteration is screaming for a `routes.stream()` with a filter on method and URI, returning an Optional. :)